extensions/rvadapter/src/main/java/com/google/android/agera/rvadapter/RepositoryPresenterCompilerStates.java

```diff
@@ -53,12 +56,60 @@
   }
 
   /**
+   * Compiler state to optionally enable fine-grained events using {@link DiffUtil}.
+   */
+  interface RPDiff<TVal, TRet> {
+
+    /**
+     * Specifies a key extractor {@link Function} to enable generating fine-grained events using
+     * {@link DiffUtil}. Two item objects are {@linkplain Callback#areItemsTheSame the same item} if
+     * their keys are {@linkplain Object#equals equal}; and they have {@linkplain
+     * Callback#areContentsTheSame the same content} if they themselves are {@linkplain
+     * Object#equals equal}. However, if the presenter is given the same instance of collection
+     * object to generate the fine-grained events, then the presenter will consider it caused by an
+     * update from an {@linkplain Builder#addAdditionalObservable additional observable}, and send a
+     * blanket item change event covering all positions.
+     *
+     * <p>Note that {@link DiffUtil} may not be suitable for all situations due to the computation
+     * complexity and that <i>the computation is likely done from the main thread</i>. If client
+     * code has better alternatives, it should directly subclass {@link RepositoryPresenter} and
+     * provide a custom implementation in {@link RepositoryPresenter#getUpdates}.
+     *
+     * @param keyForItem A function from individual item objects to objects uniquely identifying the
```

Comment:
is there a use-case for when these (itemIdForItem and keyForItem) would be different?